tpke/src/ciphertext.rs

```diff
@@ -87,18 +95,35 @@ pub fn decrypt<E: PairingEngine>(
     )]);
     decrypt_with_shared_secret(ciphertext, &s)
 }
+
 pub fn decrypt_with_shared_secret<E: PairingEngine>(
     ciphertext: &Ciphertext<E>,
     s: &E::Fqk,
 ) -> Vec<u8> {
-    let mut plaintext = ciphertext.ciphertext.to_vec();
-    let mut cipher = shared_secret_to_chacha::<E>(s);
-    cipher.apply_keystream(&mut plaintext);
+    let nonce = blinded_to_nonce::<E>(ciphertext.nonce);
+    let ciphertext = ciphertext.ciphertext.to_vec();
+
+    let cipher = shared_secret_to_chacha::<E>(s);
+    let plaintext = cipher.decrypt(&nonce, ciphertext.as_ref()).unwrap();
 
     plaintext
 }
 
-pub fn shared_secret_to_chacha<E: PairingEngine>(s: &E::Fqk) -> ChaCha20 {
+pub fn checked_decrypt_with_shared_secret<E: PairingEngine>(
+    ciphertext: &Ciphertext<E>,
+    aad: &[u8],
+    s: &E::Fqk,
+) -> Option<Vec<u8>> {
+    if check_ciphertext_validity(ciphertext, aad) {
+        Some(decrypt_with_shared_secret(ciphertext, s))
+    } else {
+        None
```

Comment:
I don't know what's the pattern in Rust, but in Python we would like to raise an exception if the validity check fails. What's the usual approach in Rust? particularly from a misuse resistance perspective.

Reply:
I agree. I was going to throw a proper exception here later on when we figure out most of the API and typed error handling: https://github.com/nucypher/ferveo/issues/12. In the meantime, I'm going to just `panic!` here, i.e. throw an untyped error with a stack trace.